File: src/com/rian/difficultycalculator/beatmap/hitobject/sliderobject/SliderHitObject.java
package com.rian.difficultycalculator.beatmap.hitobject.sliderobject;

import com.rian.difficultycalculator.beatmap.hitobject.HitObject;
import com.rian.difficultycalculator.math.Vector2;

/**
 * Represents a hit object that can be nested into a slider.
 */
public abstract class SliderHitObject extends HitObject {
    /**
     * The index of the span at which this slider hit object lies.
     */
    protected final int spanIndex;

    /**
     * The start time of the span at which this slider hit object lies, in milliseconds.
     */
    protected final double spanStartTime;

    /**
     * @param startTime              The time at which this slider hit object starts, in milliseconds.
     * @param x                      The X position of the slider hit object relative to the play field.
     * @param y                      The Y position of the slider hit object relative to the play field.
     * @param spanIndex              The index of the span at which this slider hit object lies.
     * @param spanStartTime          The start time of the span at which this slider hit object lies, in milliseconds.
     */
    public SliderHitObject(double startTime, float x, float y, int spanIndex, double spanStartTime) {
        this(startTime, new Vector2(x, y), spanIndex, spanStartTime);
    }

    /**
     * @param startTime              The time at which this hit object starts, in milliseconds.
     * @param position               The position of the hit object relative to the play field.
     * @param spanIndex              The index of the span at which this slider hit object lies.
     * @param spanStartTime          The start time of the span at which this slider hit object lies, in milliseconds.
     */
    public SliderHitObject(double startTime, Vector2 position, int spanIndex, double spanStartTime) {
        super(startTime, position);

        this.spanIndex = spanIndex;
        this.spanStartTime = spanStartTime;
    }

    /**
     * Gets the index of the span at which this slider hit object lies.
     */
    public int getSpanIndex() {
        return spanIndex;
    }

    /**
     * Gets the start time of the span at which this slider hit object lies, in milliseconds.
     */
    public double getSpanStartTime() {
        return spanStartTime;
    }
}

File: src/com/rian/osu/beatmap/hitobject/sliderobject/SliderHitObject.kt
package com.rian.osu.beatmap.hitobject.sliderobject

import com.rian.osu.beatmap.hitobject.HitObject
import com.rian.osu.beatmap.hitobject.Slider
import com.rian.osu.math.Vector2

/**
 * Represents a hit object that can be nested into a [Slider].
 */
abstract class SliderHitObject(
    /**
     * The time at which this [SliderHitObject] starts, in milliseconds.
     */
    startTime: Double,

    /**
     * The position of this [SliderHitObject] relative to the play field.
     */
    position: Vector2,

    /**
     * The index of the span at which this [SliderHitObject] lies.
     */
    @JvmField
    val spanIndex: Int,

    /**
     * The start time of the span at which this [SliderHitObject] lies, in milliseconds.
     */
    @JvmField
    val spanStartTime: Double
) : HitObject(startTime, position, false, 0)
